### PR TITLE
Fix #116: Option --space-after-function-name doesn't work with function pointer syntax.

### LINKED ISSUE #116: Option --space-after-function-name doesn't work with function pointer syntax
I think it's a recent change.
```c
int foo(void);

typedef int (*Fun0)(void);
typedef int (*Fun1)(int (*param)(void));

struct Foo {
  int (*bar)(void);
};
```
when run with `--space-after-function-name=false`, gives:
```d
int foo();

alias Fun0 = int function ();
alias Fun1 = int function (int function () param);

struct Foo
{
    int function () bar;
}
```

### PR DESCRIPTION
**Pull request by ciechowoj**
As in the title.

### CODE REVIEW

**Comment by jacob-carlborg** [comment]
Update the changelog please.

**Comment by ciechowoj** [comment]
Done. However some tests don't pass due to [this](https://github.com/jacob-carlborg/dstep/pull/118/files#diff-0c0cbcae56297a371672c36f64de91f3R101), so this PR will have to wait until #118 is merged.

**Comment by ciechowoj** [comment]
@jacob-carlborg  Rebased and ready to merge.

File: unit_tests/UnitTests.d [hunk 1]
```diff
@@ -905,3 +905,34 @@ void foo ();
 D", options));
 
 }
+
+// Put or put not space after 'function' keyword when
```
Comment (jacob-carlborg) [inline unit_tests/UnitTests.d]: `put not` -> `don't put`.

File: unit_tests/issues/Issue116.d [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+import Common;
+import dstep.translator.Options;
+
+// Fix 116: Option --space-after-function-name doesn't work with function
+// pointer syntax.
+unittest
+{
+    Options options;
+  	options.spaceAfterFunctionName = false;
```
Comment (jacob-carlborg) [inline unit_tests/issues/Issue116.d]: Indentation is not correct.

File: unit_tests/UnitTests.d (hunk 1 above)
Comment (ciechowoj) [inline unit_tests/UnitTests.d]: "Do or do not, there is no try." ~ Yoda : D

Comment (ciechowoj) [inline unit_tests/UnitTests.d]: Fixed, anyway.

File: unit_tests/issues/Issue116.d (hunk 2 above)
Comment (ciechowoj) [inline unit_tests/issues/Issue116.d]: Fixed.

**Comment by jacob-carlborg** [comment]
master is failing, so I'm going to wait with accepting this PR.

**Comment by ciechowoj** [comment]
Sure.

**Comment by ciechowoj** [comment]
I've added here an another commit that is supposed to fix the problem with AppVeyor, I've just mentioned [here](https://github.com/jacob-carlborg/dstep/pull/113#issuecomment-279396135).

**Comment by jacob-carlborg** [comment]
I'm running the tests again.

**Comment by ciechowoj** [comment]
There is something wrong with AppVeyor x86_64 build. I think it has not been passing since quite a long, however it haven't been reported. Check this out [link](https://ci.appveyor.com/project/jacob-carlborg/dstep/build/1.0.169/job/bh60xfdnc8suwdt5). The CI is green despite the fact there were errors.

**Comment by jacob-carlborg** [comment]
Yeah. This seems to be the first time it started to fail: https://ci.appveyor.com/project/jacob-carlborg/dstep/build/1.0.85/job/v1xcjkuk22e5oe83.

Is it necessary to manually set the exist code to have the test fail? Seems pretty bad.

**Comment by ciechowoj** [comment]
> Is it necessary to manually set the exist code to have the test fail? Seems pretty bad.

I see this as mapping of one error codes to some other error codes. But I'm not sure, why it doesn't work. I'll investigate it, when I have some spare time.

**Comment by jacob-carlborg** [comment]
> I see this as mapping of one error codes to some other error codes. But I'm not sure, why it doesn't work. Maybe there is a better solution. I'll investigate it, when I have some spare time.

As long as it properly fails the test I'm satisfied. But now we need to figure out why the 64bit is failing.

**Comment by ciechowoj** [comment]
Ok.

**Comment by ciechowoj** [comment]
I've tried to fix that windows build and there are some problems. The main reason the x64 tests was a single unit test I commented it out because I haven't figured out how to fix it yet. 

However there were issues with `time_t` and adding the `std.time` import. I do not know what had made the handling of time_t to be broken, but handling of includes and imports needs an improvement.

**Comment by jacob-carlborg** [comment]
> I've tried to fix that windows build and there are some problems. The main reason the x64 tests were failing was a single unit test. I commented it out because I haven't figured out how to fix it yet. There is something wrong with throwing an exception in throwCollisionError in Context.d:207...

Ok. Do you expect it to be a logical error, or could it be some kind of bug with the exception handling in the compiler?

How do you recommend we proceed, just comment out that test for now and move on, or? 

> What is more, there were issues with time_t and adding the std.time import. I do not know what had made the handling of time_t to be broken, but handling of includes and imports needs an improvement. I've added a quick and dirty fix but it needs to be changed

I'm not sure if this is the same issue but I've noticed (at least in the header files on macOS) that a lot of the symbols from the standard library (or Posix) are not defined in the headers the documentation says to include. Rather they're defined in some platform/system specific files, which are then included in the header files the documentation says to include.

**Comment by ciechowoj** [comment]
> Ok. Do you expect it to be a logical error, or could it be some kind of bug with the exception handling in the compiler?

Currently, almost no idea. It seems to work with 3.7 and break with 3.8, maybe some .dll incompatibility?

> How do you recommend we proceed, just comment out that test for now and move on, or?

I do not have more spare time to spend on it at the moment. The unit test in question tests the `--collision-action=abort" feature which isn't that critical. I would leave the test commented and open a corresponding issue on github. 

I am going to fix it eventually...

> I'm not sure if this is the same issue but I've noticed (at least in the header files on macOS) that a lot of the symbols from the standard library (or Posix) are not defined in the headers the documentation says to include. Rather they're defined in some platform/system specific files, which are then included in the header files the documentation says to include.

Yeah, that is the exact source of problem.

**Comment by jacob-carlborg** [comment]
> Yeah, that is the exact source of problem.

What if we extend the list of include mappings to also include the name of the type/symbol. Something like:

```d
auto knownIncludes = [Foo("time_t", `sys\types.h`) : "core.stdc.time"];
```

Need to come up with a better name for `Foo` of course.

**Comment by ciechowoj** [comment]
> What if we extend the list of include mappings to also include the name of the type/symbol. Something like:
> Need to come up with a better name for Foo of course.

Yes, that seems like a viable solution for the time being .

**Comment by jacob-carlborg** [comment]
@ciechowoj is there a possibility that you could extract the fix for the failing Windows tests as a separate pull requests? I've been trying for the last couple of days to setup an environment for compiling DStep on Windows but have failed.

**Comment by ciechowoj** [comment]
> @ciechowoj is there a possibility that you could extract the fix for the failing Windows tests as a separate pull requests? 

Yeah, I'll do it tomorrow, if I don't forget. 

> I've been trying for the last couple of days to setup an environment for compiling DStep on Windows but have failed so far.

Hmm, it is quite straightforward for x86_64, at least I created the setup that can compile dstep within 1h on virtual machine. You need git, dmd, dub and LLVM, all of them have windows installers...

**Comment by jacob-carlborg** [comment]
> Hmm, it is quite straightforward for x86_64, at least I created the setup that can compile dstep within 1h on virtual machine. You need git, dmd, dub and LLVM, all of them have windows installers...

Most of the issues have been with Visual Studio. Currently I'm running inside the Visual Studio shell (or whatever it's called) and DStep (libclang) cannot find the C standard library header files. I hvae basically no experience with developing on Windows.

**Comment by jacob-carlborg** [comment]
> Yeah, I'll do it tomorrow, if I don't forget.

Thanks, no rush.

**Comment by ciechowoj** [comment]
> Most of the issues have been with Visual Studio. Currently I'm running inside the Visual Studio shell (or whatever it's called) and DStep (libclang) cannot find the C standard library header files. I hvae basically no experience with developing on Windows.

Oh, I see. Truth say, I haven't made it to that state as well, I only made it to compile, so that I can find out why the tests fail on AppVeyor. They fail due to lack of standard headers on my local VirtualBox installation as well. AppVeyor seems to use headers from its MinGW installation. Soon, I should have access to the computer with normal Windows installation, so I will take a closer look at that issue.

**Comment by jacob-carlborg** [comment]
@ciechowoj BTW, have you seen this [1]? It's possible to remotely connect to a failed build.

[1] https://www.appveyor.com/docs/how-to/rdp-to-build-worker/

**Comment by ciechowoj** [comment]
No, I haven't. I could be useful, indeed.

**Comment by ciechowoj** [comment]
@jacob-carlborg 

> > Yeah, I'll do it tomorrow, if I don't forget.
> Thanks, no rush.

Done.

**Comment by jacob-carlborg** [comment]
👌

**Comment by jacob-carlborg** [comment]
Thanks.

**Pull request merged**